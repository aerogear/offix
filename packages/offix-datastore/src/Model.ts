import { Storage, StoreChangeEvent, EventTypes } from "./storage";
import { createPredicate, Predicate } from "./predicates";

export interface FieldOptions {
    /** GraphQL type */
    type: string;
    /** GraphQL key */
    key: string;
    // TODO
    format?: {  };
}

/**
 * Defines the properties expected in the Fields object for a model
 */
export type Fields<T> = {
    [P in keyof T]: FieldOptions
};

/**
 * Provides CRUD capabilities for a model
 */
export class Model<T> {
    private storeName: string;
    private fields: Fields<T>;
    private getStorage: () => Storage;

    constructor(
        storeName: string,
        fields: Fields<T>,
        getStorage: () => Storage
    ) {
        this.storeName = storeName;
        this.fields = fields;
        this.getStorage = getStorage;
    }

    public save(input: T): Promise<T> {
        return this.getStorage().save(this.storeName, input);
    }

    public query(predicateFunction?: Predicate<T>) {
        if (!predicateFunction) {return this.getStorage().query(this.storeName);}

        const modelPredicate = createPredicate(this.fields);
        const predicate = predicateFunction(modelPredicate);
        return this.getStorage().query(this.storeName, predicate);
    }

    public update(input: Partial<T>, predicateFunction?: Predicate<T>) {
        if (!predicateFunction) {
            return this.getStorage().update(this.storeName, input);
        }

        const modelPredicate = createPredicate(this.fields);
        const predicate = predicateFunction(modelPredicate);
        return this.getStorage().update(this.storeName, input, predicate);
    }

    public remove(predicateFunction?: Predicate<T>) {
        if (!predicateFunction) {return this.getStorage().remove(this.storeName);}

        const modelPredicate = createPredicate(this.fields);
        const predicate = predicateFunction(modelPredicate);
        return this.getStorage().remove(this.storeName, predicate);
    }

    // TODO fix event listening
    public on(event: EventTypes, listener: (event: StoreChangeEvent) => void) {
        return this.getStorage().storeChangeEventStream.subscribe(listener);
    }
}
