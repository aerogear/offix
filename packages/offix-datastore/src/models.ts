import { GraphQLSchema, GraphQLObjectType } from "graphql";
import { parseMarker } from "graphql-metadata";
import { getUserTypesFromSchema } from "@graphql-toolkit/common";

export interface Model {
    __typename: string;
}

export interface PersistedModel extends Model {
    id: string;
}

export function extractModelsFromSchema(schema: GraphQLSchema): Model[] {
    const types = getUserTypesFromSchema(schema);
    const userTypes = types.filter((modelType: GraphQLObjectType) => parseMarker("model", modelType.description));
    return userTypes.map((value) => {
        return {
            __typename: value.name
        };
    });
}
