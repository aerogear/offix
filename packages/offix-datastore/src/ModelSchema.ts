import invariant from "tiny-invariant";
import { JSONSchema7 } from "json-schema";

export interface DataSyncProperties extends JSONSchema7 {
  index?: boolean;
  primary?: boolean;
  default?: any;
  encrypted?: boolean;
  // TODO investigate scalars
}

export declare class DataSyncJsonSchema<T> {
  name: string;
  namespace?: string;
  version: number;
  indexes?: string[];
  encrypted?: string[];
  primaryKey?: string;
  type: "object";
  properties?: {
    [key: string]: DataSyncProperties;
  };
};

export class ModelSchema<T = any>{
  private name: string;
  private namespace: string;
  private primaryKey: string;
  private fields: string[];
  private encrypted: string[];
  private version: number;
  private indexes: string[] = [];

  constructor(schema: DataSyncJsonSchema<T>) {
    invariant(schema, "Schema cannot be undefined");
    this.version = schema.version || 0;
    this.name = schema.name;
    this.namespace = schema.namespace || "user";
    this.fields = extractFields(schema);
    this.primaryKey = extractPrimary(schema);
    this.indexes = extractIndexes(schema);
    this.encrypted = extractEncryptedFields(schema);
  }

  public fill(): void {
    // TODO fill object with default values
  }

  public validate() {
    // TODO validate the schema
  }

  public getName() {
    return this.name;
  }

  public getNamespace() {
    return this.namespace;
  }

  public getIndexes(): string[] {
    return this.indexes;
  }

  public getPrimaryKey(): string {
    return this.primaryKey;
  }

  public getFields(): string[] {
    return this.fields;
  }

  public getVersion(): number {
    return this.version;
  }

  public getEncryptedFields(): string[] {
    return this.encrypted;
  }

};

function extractFields<T = any>(schema: DataSyncJsonSchema<T>): string[] {
  invariant(schema.properties, "Properties cannot be undefined");
  return Object.keys(schema.properties);
}

function extractIndexes<T = any>(schema: DataSyncJsonSchema<T>) {
  const { properties, indexes } = schema;
  if (!indexes) {
    invariant(properties, "Schema is undefined");
    return Object.keys(properties)
      .filter((key) => {
        if (properties[key].index) {
          return key;
        };
      });
  }
  return indexes;
}

function extractPrimary<T = any>(schema: DataSyncJsonSchema<T>): string {
  const { properties, primaryKey } = schema;
  if (!primaryKey) {
    invariant(properties, "Schema is undefined");
    const obj = Object.keys(properties)
      .find((key) => properties[key].primary);
    return obj || "_id";
  }
  return primaryKey;
}

function extractEncryptedFields<T = any>(schema: DataSyncJsonSchema<T>): string[] {
  const { properties, encrypted } = schema;
  if (!encrypted) {
    invariant(properties, "Schema is undefined");
    return Object.keys(properties)
      .filter((key) => properties[key].encrypted);
  }
  return encrypted;
}

export function createModelSchema<T = any>(schema: DataSyncJsonSchema<T>): ModelSchema<T> {
  const modelSchema =  new ModelSchema<T>(schema);
  // TODO validation could potentially be run
  // in the constructor
  modelSchema.validate();
  return modelSchema;
}
