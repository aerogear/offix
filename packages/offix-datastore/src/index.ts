export { CRUDEvents } from "./storage";
export * from "./DataStore";
