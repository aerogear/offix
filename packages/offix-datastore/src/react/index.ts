export * from "./hooks/query";
export * from "./hooks/save";
export * from "./hooks/update";
export * from "./hooks/delete";
