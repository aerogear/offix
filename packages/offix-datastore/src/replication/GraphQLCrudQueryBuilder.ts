import gql from "graphql-tag";

import { GraphQLQueryBuilder, GraphQLQueries } from "./GraphQLReplicator";
import { Model } from "../Model";

/**
 * Builds GraphQL queries for models following
 * the [GraphQLCRUD specification]{@link https://graphqlcrud.org/}
 */
export class GraphQLCrudQueryBuilder implements GraphQLQueryBuilder {
    build(models: Model<any>[]): Map<string, GraphQLQueries> {
        const queriesMap: Map<string, GraphQLQueries> = new Map();

        models.forEach((model) => {
            const fields = model.getFields();
            const fieldsBuilder: string[] = Object.keys(fields).map((key) => {
                const graphQLKey = fields[key].key;
                return graphQLKey;
            });
            const graphQLFields = fieldsBuilder.join("\n");
            const modelName = model.getName();

            const mutations = {
                create: gql`
                mutation create${modelName}($input: Create${modelName}Input!) {
                    create${modelName}(input: $input) {
                        ${graphQLFields}
                    }
                }`,
                update: gql`
                mutation update${modelName}($input: Mutate${modelName}Input!) {
                    update${modelName}(input: $input) {
                        ${graphQLFields}
                    }
                }`,
                delete: gql`
                mutation delete${modelName}($input: Mutate${modelName}Input!) {
                    delete${modelName}(input: $input) {
                        ${graphQLFields}
                    }
                }`
            };

            queriesMap.set(model.getStoreName(), { mutations });
        });

        return queriesMap;
    }
}
