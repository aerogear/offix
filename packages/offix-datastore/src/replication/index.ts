export * from "./ReplicationEngine";
export * from "./Replicator";
export * from "./GraphQLReplicator";
export * from "./GraphQLCrudQueryBuilder";
