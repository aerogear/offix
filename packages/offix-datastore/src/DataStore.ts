import { Storage } from "./storage";
import { Model, Fields } from "./Model";
import { ReplicationEngine, GraphQLReplicator, UrqlGraphQLClient, GraphQLCrudQueryBuilder } from "./replication";

/**
 * Configuration Options for DataStore
 */
export interface DataStoreConfig {
    /**
     * The Database name
     */
    dbName: string;

    /**
     * The GraphQL endpoint for synchronisation
     */
    url: string;

    /**
     * The Schema Version number. Used to trigger a Schema upgrade
     */
    schemaVersion?: number;
}

export class DataStore {
    private dbName: string;
    private schemaVersion: number;
    private models: Model<any>[];
    private storage?: Storage;
    private url: string;

    constructor(config: DataStoreConfig) {
        this.dbName = config.dbName;
        this.schemaVersion = config.schemaVersion || 1; // return 1 is schemaVersion is undefined or 0
        this.url = config.url;
        this.models = [];
    }

    public create<T>(name: string, storeName: string, fields: Fields<T>) {
        const model = new Model<T>(name, storeName, fields, () => {
            if (this.storage) { return this.storage; }
            throw new Error("DataStore has not been initialised");
        });
        this.models.push(model);
        return model;
    }

    public init() {
        this.storage = new Storage(this.dbName, this.models, this.schemaVersion);

        const gqlClient = new UrqlGraphQLClient(this.url);
        const queryBuilder = new GraphQLCrudQueryBuilder();
        const queries = queryBuilder.build(this.models);
        const gqlReplicator = new GraphQLReplicator(gqlClient, queries);

        const engine = new ReplicationEngine(gqlReplicator, (this.storage as Storage));
        engine.start();
    }
}
