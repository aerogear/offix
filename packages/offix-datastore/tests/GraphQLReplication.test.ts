import { GraphQLCrudQueryBuilder } from "../src/replication";
import { DataStore } from "../src/DataStore";

test("Test Query generation", () => {
    const dataStore = new DataStore("test");
    const model = dataStore.create<any>("Test", "test", {
        id: {
            type: "ID",
            key: "id"
        },
        title: {
            type: "String",
            key: "title"
        }
    });
    const queryBuilder = new GraphQLCrudQueryBuilder();
    const queries = queryBuilder.build([model]);
    const modelQueries = queries.get(model.getName());

    expect(modelQueries).toMatchSnapshot();
});
