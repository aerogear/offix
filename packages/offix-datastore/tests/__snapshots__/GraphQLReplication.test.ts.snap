// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Query generation 1`] = `
Object {
  "mutations": Object {
    "create": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "createTest",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "input",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "input",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "createTest",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "title",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "CreateTestInput",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "input",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 185,
        "start": 0,
      },
    },
    "delete": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "deleteTest",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "input",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "input",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "deleteTest",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "title",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "MutateTestInput",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "input",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 185,
        "start": 0,
      },
    },
    "update": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "updateTest",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "input",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "input",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "updateTest",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "title",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "MutateTestInput",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "input",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 185,
        "start": 0,
      },
    },
  },
}
`;
